src/codeActionProvider.ts

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import * as path from 'path';
+import { CancellationToken, CodeAction, CodeActionContext, CodeActionKind, CodeActionProvider, Range, Selection, TextDocument } from 'vscode';
+import { getSourcePaths } from './utils/commandUtils';
+
+export class TestCodeActionProvider implements CodeActionProvider {
+
+    private sourcePaths: ISourcePath[];
+
+    public async provideCodeActions(document: TextDocument, range: Range | Selection, _context: CodeActionContext, token: CancellationToken): Promise<CodeAction[] | null> {
+        if (!this.sourcePaths) {
+            this.sourcePaths = await getSourcePaths();
```

Comment:
No need to restrict it to files under source paths. If the source paths have changed, it won't work.

You can just enable the source action at any *.java file.

Reply:
And will lessen the time to resolve the code actions. Makes sense.